libass/ass_render.c

```diff
@@ -2203,244 +2134,192 @@ static void render_and_combine_glyphs(ASS_Renderer *render_priv,
-                    text_info->combined_bitmaps = combined_info =
-                        realloc(combined_info,
-                                sizeof(CombinedBitmapInfo) * text_info->max_bitmaps);
+                    size_t new_size = 2 * text_info->max_bitmaps;
+                    if (!ASS_REALLOC_ARRAY(text_info->combined_bitmaps, new_size))
+                        continue;
+                    text_info->max_bitmaps = new_size;
+                    combined_info = text_info->combined_bitmaps;
                 }
-
-                current_info = &combined_info[nb_bitmaps - 1];
-
-                current_info->pos.x = min_bm_x;
-                current_info->pos.y = min_bm_y;
+                current_info = &combined_info[nb_bitmaps];
 
                 current_info->first_pos_x = info->bbox.xMax >> 6;
 
                 memcpy(&current_info->c, &info->c, sizeof(info->c));
                 current_info->effect_type = info->effect_type;
                 current_info->effect_timing = info->effect_timing;
-                current_info->be = info->be;
-                current_info->blur = info->blur;
-                current_info->shadow_x = info->shadow_x;
-                current_info->shadow_y = info->shadow_y;
-                current_info->frx = info->frx;
-                current_info->fry = info->fry;
-                current_info->frz = info->frz;
-                current_info->fax = info->fax;
-                current_info->fay = info->fay;
-                current_info->scale_x = info->scale_x;
-                current_info->scale_y = info->scale_y;
-                current_info->border_style = info->border_style;
-                current_info->border_x = info->border_x;
-                current_info->border_y = info->border_y;
-                current_info->hspacing = info->hspacing;
-                current_info->italic = info->italic;
-                current_info->bold = info->bold;
-                current_info->flags = info->flags;
-
-                current_info->shift_x = key->shift_x;
-                current_info->shift_y = key->shift_y;
-                current_info->advance = key->advance;
-
-                current_info->has_border = !!info->border;
-
-                current_info->has_outline = 0;
-                current_info->cached = 0;
-                current_info->is_drawing = 0;
-
+                current_info->filter.be = info->be;
+                current_info->filter.blur = 2 * info->blur * render_priv->blur_scale;
+                current_info->filter.shadow.x = double_to_d6(info->shadow_x * render_priv->border_scale);
+                current_info->filter.shadow.y = double_to_d6(info->shadow_y * render_priv->border_scale);
+                current_info->filter.border_style = info->border_style;
+                current_info->filter.nonzero_border = info->border_x || info->border_y;
+                current_info->filter.show_shadow = info->shadow_x || info->shadow_y;
```

Comment:
This is needed to ensure we show a shadow even if `shadow.x` or `shadow.y` underflow to zero because of the scaling, right? (And it matters because the glyphs can be translucent, making even a zero-offset shadow visible.)

Don’t you think it would be better to rename this to `nonzero_shadow`? This would add consistency with `nonzero_border` and reduce confusion with `draw_shadow`.